src/rule.rs

```diff
@@ -0,0 +1,736 @@
+        }
+    }
+
+    pub(crate) fn check_call_names<Pat>(self, cx: &Context<Pat>, grammar: &crate::Grammar) {
+        match cx[self] {
+            Rule::Empty | Rule::Eat(_) => {}
+            Rule::Call(rule) => {
+                assert!(
+                    grammar.rules.contains_key(&rule),
+                    "no rule named `{}`",
+                    cx[rule]
+                );
+            }
+            Rule::Concat([left, right]) => {
+                left.check_call_names(cx, grammar);
+                right.check_call_names(cx, grammar);
+            }
+            Rule::Or(ref rules) => {
+                for rule in rules {
+                    rule.check_call_names(cx, grammar);
+                }
+            }
+            Rule::Opt(rule) => rule.check_call_names(cx, grammar),
+            Rule::RepeatMany(elem, sep) | Rule::RepeatMore(elem, sep) => {
+                elem.check_call_names(cx, grammar);
+                if let Some((sep, _)) = sep {
+                    sep.check_call_names(cx, grammar);
+                }
+            }
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum MaybeKnown<T> {
+    Known(T),
+    Unknown,
+}
+
+impl BitOr for MaybeKnown<bool> {
+    type Output = Self;
+
+    fn bitor(self, rhs: Self) -> Self {
+        match (self, rhs) {
+            (MaybeKnown::Known(true), _) | (_, MaybeKnown::Known(true)) => MaybeKnown::Known(true),
+            (MaybeKnown::Known(false), x) | (x, MaybeKnown::Known(false)) => x,
+            (MaybeKnown::Unknown, MaybeKnown::Unknown) => MaybeKnown::Unknown,
+        }
+    }
+}
+
+impl BitAnd for MaybeKnown<bool> {
+    type Output = Self;
+
+    fn bitand(self, rhs: Self) -> Self {
+        match (self, rhs) {
+            (MaybeKnown::Known(false), _) | (_, MaybeKnown::Known(false)) => {
+                MaybeKnown::Known(false)
+            }
+            (MaybeKnown::Known(true), x) | (x, MaybeKnown::Known(true)) => x,
+            (MaybeKnown::Unknown, MaybeKnown::Unknown) => MaybeKnown::Unknown,
+        }
+    }
+}
+
+pub trait MatchesEmpty {
+    fn matches_empty(&self) -> MaybeKnown<bool>;
+}
+
+pub trait Folder<Pat: Eq + Hash>: Sized {
+    fn cx(&mut self) -> &mut Context<Pat>;
+    fn fold_leaf(&mut self, rule: RuleWithNamedFields) -> RuleWithNamedFields {
+        rule
+    }
+    fn fold_concat(
+        &mut self,
+        left: RuleWithNamedFields,
+        right: RuleWithNamedFields,
+    ) -> RuleWithNamedFields {
+        (left.fold(self) + right.fold(self)).finish(self.cx())
+    }
+    fn fold_or(
+        &mut self,
+        mut rules: impl Iterator<Item = RuleWithNamedFields>,
+    ) -> RuleWithNamedFields {
+        let first = rules.next().unwrap().fold(self);
+        rules.fold(first, |or, rule| (or | rule.fold(self)).finish(self.cx()))
+    }
+    fn fold_opt(&mut self, rule: RuleWithNamedFields) -> RuleWithNamedFields {
+        rule.fold(self).opt().finish(self.cx())
+    }
+    fn fold_repeat_many(
```

Comment:
You split these into `_sep` variants above, would it make sense to do so here as well?

Reply:
That was only because I didn't want to add another trait, I think keeping the orthogonality in the `fold` is fine.